Sources/Apollo/ApolloInterceptorReentrantWrapper.swift

```diff
@@ -0,0 +1,109 @@
+#if !COCOAPODS
+import ApolloAPI
+#endif
+
+/// A custom implementation of RequestChain that wraps an ApolloInterceptor instance to provide
+/// re-entrant behaviour to the request chain. This is required for the request chain to support
+/// network flows such as GraphQL Subscriptions where interceptors can call back into the request
+/// chain multiple times.
+class ApolloInterceptorReentrantWrapper: RequestChain {
+  @Atomic var isCancelled: Bool = false
+
+  let wrappedInterceptor: ApolloInterceptor
+  let requestChain: Unmanaged<InterceptorRequestChain>
+  let index: Int
+
+  init(
+    interceptor: ApolloInterceptor,
+    requestChain: Unmanaged<InterceptorRequestChain>,
```

Comment:
Love that we are using `Unmanaged` here!

Reply:
Do we have a unit test to ensure we aren't accidentally creating a zombie when using this?

Reply:
The watcher tests we fixed for the other day are an indirect way of testing of for a retain cycle but I added a more direct test on the request chain itself - https://github.com/apollographql/apollo-ios/pull/2870/commits/21857e9b8b823ab18179398dfe96f58fca8b5bbb